server2/Arcadia.Assistant/Arcadia.Assistant.Notifications/Notifications.cs

```diff
@@ -0,0 +1,149 @@
+
+    using UserPreferences.Contracts;
+
+    /// <summary>
+    ///     An instance of this class is created for each service instance by the Service Fabric runtime.
+    /// </summary>
+    public class Notifications : StatelessService, INotifications
+    {
+        private readonly IDeviceRegistry deviceRegistry;
+        private readonly ILogger logger;
+
+        private readonly Dictionary<string, IEnumerable<NotificationType>> notificationProvidersMap;
+        private readonly INotificationSettings notificationSettings;
+        private readonly IPushNotifications pushNotifications;
+        private readonly IUsersPreferencesStorage userPreferences;
+
+        public Notifications(
+            StatelessServiceContext context,
+            INotificationSettings notificationSettings,
+            IDeviceRegistry deviceRegistry,
+            IUsersPreferencesStorage userPreferences,
+            IPushNotifications pushNotifications,
+            ILogger<Notifications> logger)
+            : base(context)
+        {
+            this.notificationSettings = notificationSettings;
+            this.deviceRegistry = deviceRegistry;
+            this.userPreferences = userPreferences;
+            this.pushNotifications = pushNotifications;
+            this.logger = logger;
+            this.notificationProvidersMap = this.notificationSettings.ClientNotificationProvidersMap;
+        }
+
+        public async Task Send(IEnumerable<EmployeeId> employeeId, NotificationMessage notificationMessage, CancellationToken cancellationToken)
+        {
+            if (employeeId.IsNullOrEmpty())
+            {
+                return;
+            }
+
+            var notificationProviders = this.notificationProvidersMap.TryGetValue(notificationMessage.ClientName, out var map) ? map : Enum.GetValues(typeof(NotificationType)).OfType<NotificationType>().ToList();
+            foreach (var providerType in notificationProviders)
+            {
+                switch (providerType)
+                {
+                    case NotificationType.Push:
+                        if (this.notificationSettings.EnablePush)
+                        {
+                            var tokens = await this.GetDeviceTokens(employeeId, cancellationToken);
+                            await this.SendPushNotification(tokens, notificationMessage, cancellationToken);
+                        }
+
+                        break;
+                }
+            }
+        }
+
+        private async Task SendPushNotification(Dictionary<EmployeeId, IEnumerable<DeviceRegistryEntry>> deviceTokens, NotificationMessage notificationMessage, CancellationToken cancellationToken)
+        {
+            var notificationContent = new PushNotificationContent
+            {
+                Title = notificationMessage.Subject,
+                Body = notificationMessage.ShortText
+            };
+
+            foreach (var employeeId in deviceTokens.Keys)
```

Comment:
you could have iterated over Values instead of accessing [] each time in a loop

Reply:
Done.